### PR TITLE
feat(env): Google::Cloud::Env provides additional configuration options including retries and cache control

### PR DESCRIPTION
**Pull request by dazuma**
Some steps towards addressing #3125

### CODE REVIEW

File: google-cloud-env/lib/google/cloud/env.rb [hunk 1]
```diff
@@ -65,22 +65,55 @@ class Env
...
+      #     parameter, and should add middleware and set the adapter as if it
+      #     were passed as a block to `Faraday.new`.
       # @param [Faraday::Connection] connection Faraday connection to use.
-      # @param [Hash] metadata_cache Mock cache.
+      #     If specified, overrides the `request_timeout`, `retries`, and
+      #     `build_proc` settings.
       #
-      def initialize env: nil, connection: nil, metadata_cache: nil
+      def initialize env: nil, connection: nil, metadata_cache: nil,
+                     request_timeout: 0.1, retries: 0, build_proc: nil
```
Comment (blowmage) [inline google-cloud-env/lib/google/cloud/env.rb]: How will these defaults be set on the `Google::Cloud.env` instance?

Comment (dazuma) [reply google-cloud-env/lib/google/cloud/env.rb]: Yeah, good question. They're currently set to match the previous behavior (100ms timeout, and no retry) but perhaps it should default to something else. I'll try setting it to one retry. (We have an internal App Engine prober that seems to be failing intermittently on metadata server queries, so we'll have a real world test case to measure the effect.)

File: google-cloud-env/lib/google/cloud/env.rb [hunk 2]
```diff
@@ -353,7 +351,7 @@ def metadata?
           rescue *METADATA_FAILURE_EXCEPTIONS
             retries_remaining -= 1
             if retries_remaining >= 0
-              sleep @retry_interval if @retry_interval
+              sleep @retry_interval
```
Comment (blowmage) [inline google-cloud-env/lib/google/cloud/env.rb]: A simple delay and retry is sufficent? An incremental backoff is not required?

Comment (dazuma) [reply google-cloud-env/lib/google/cloud/env.rb]: I was going to say I didn't think it was important... but I guess it's not hard to put in.

**Comment by quartzmo** [comment]
@dazuma You can rebase on master to fix this build failure, which is in storage unit tests.

**Pull request merged**